### PR TITLE
use previous epoch rent collector in expected rent collection

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem

Fix bug in expected rent collection related to rent paying accounts where the account's rent slot hasn't occurred in the current epoch yet. We need to use a rent collector that is one epoch older to correctly calculate rent if the slot in this epoch has not been hit yet.

#### Summary of Changes

Fixes #

### CODE REVIEW

File: runtime/src/expected_rent_collection.rs [hunk 1]
```diff
@@ -499,22 +504,28 @@ impl ExpectedRentCollection {
             return None;
         }
 
+        let rent_collector_previous;
+        if use_previous_epoch_rent_collector {
+            // keep in mind the storage slot could be 0..inf epochs in the past
+            // we want to swap the rent collector for one whose epoch is the previous epoch
+            let mut rent_collector_temp = rent_collector.clone();
+            rent_collector_temp.epoch = rent_collector.epoch.saturating_sub(1); // previous epoch
```
[the remark this answers is not in the document]
Comment (jeffwashington) [reply runtime/src/expected_rent_collection.rs:512]: We are asked to figure out what this account should look like as of slot `max_slot_in_storages_inclusive`.
`max_slot_in_storages_inclusive` is in epoch `N`. We have a rent collector that is in the same epoch, `N`.
If the rent collection slot for this account has not occurred in epoch `N` yet, then the rent collector that is applicable to this account has epoch `N-1`. So, this defines what rent collector to use. The account itself may not have been written to disk for say 10 epochs. It is important that we use a rent collector of the correct epoch as of the last rent collection slot for this account. The rent collector figures out that it has to collect rent or whatever for the past 10 slots. This change make sure the rent collector has the correct epoch, which can either be `N` (which it incorrectly always was before) or `N-1` if we haven't hit the rent collection slot yet.

[the remark this answers is not in the document]
Comment (jeffwashington) [reply runtime/src/expected_rent_collection.rs:512]: We were actually fixing up the rent epoch (-1) previously, but that turned out to be incorrect for a rent-paying account where we hadn't hit the rent collection slot in the current epoch yet.

[the remark this answers is not in the document]
Comment (jeffwashington) [reply runtime/src/expected_rent_collection.rs:512]: now we use the correct epoch in all cases.

**Review by brooksprumo — APPROVED** [review]
lgtm

**Pull request merged**